compiler/rustc_mir/src/transform/inline.rs

```diff
@@ -216,19 +212,18 @@ impl Inliner<'tcx> {
             }
 
             // Avoid a cycle here by only using `instance_mir` only if we have
-            // a lower `HirId` than the callee. This ensures that the callee will
-            // not inline us. This trick only works without incremental compilation.
-            // So don't do it if that is enabled.
-            if !self.tcx.dep_graph.is_fully_enabled() && self.hir_id < callee_hir_id {
+            // a lower `DefPathHash` than the callee. This ensures that the callee will
+            // not inline us. This trick only works even with incremental compilation,
+            // since `DefPathHash` is stable.
+            if self.tcx.def_path_hash(caller_def_id)
+                < self.tcx.def_path_hash(callee_def_id.to_def_id())
+            {
                 return Ok(());
             }
```

Comment:
Why does this still exist now that we have `mir_callgraph_reachable`?

Reply:
Two advantages:
1. it halves the amount of calls to `mir_callgraph_reachable`,
2. it allows for inlining even in the presence of a cycle.

Reply:
But this completely omits half of all inlining candidates. Also with hir id comparisons this would give a consistent result. Now any rustc version increment would cause the result to change as the def path hash contains the `-Cmetadata` calculated by cargo based on the compiler version, which would break the tests and seemingly randomly regress/improve performance when enabling mir inlining.

Reply:
> But this completely omits half of all inlining candidates.

To the contrary, this branch *allows* half of all candidates without further examination.

> Also with hir id comparisons this would give a consistent result. Now any rustc version increment would cause the result to change as the def path hash contains the -Cmetadata calculated by cargo based on the compiler version, which would break the tests and seemingly randomly regress/improve performance when enabling mir inlining.

I will change the test to only consider the local part of the DefPathHash, removing the influence of -Cmetadata.

Reply:
I would prefer to avoid optimization decisions being nondeterministic or otherwise uniform - we already see a lot of pain due to CGU partitioning, for example. This seems likely to be even less explainable, right?

Reply:
The `mir_callgraph_reachable` is relatively expensive. If we can avoid calling it half the time that seems like a nice win!

The unpredictable nature of the check is a disadvantage, but the condition is effectively: `caller_hash < callee_hash || !reachable(callee, caller)`, which means that it becomes unpredictable only in a presence of a cycle, since otherwise MIR would be available for inlining anyway. I think this is a reasonable trade-off to make.

BTW. Does comparing `local_hash` help? It seems to incorporate complete `DefPathHash` of a parent: https://github.com/rust-lang/rust/blob/fa72878a61f2b0a2127fe7d700724642fc79ec66/compiler/rustc_hir/src/definitions.rs#L143-L156

Reply:
> which means that it becomes unpredictable only in a presence of a cycle, since otherwise MIR would be available for inlining anyway. I think this is a reasonable trade-off to make.

I see. In that case I think it is fine.